twenty-first/src/util_types/mmr/mmr_trait.rs

```diff
@@ -39,42 +39,129 @@ impl LeafMutation {
+///     ╱    ╲        ╱    ╲        ╱    ╲        │
+///    2      5      9      12     17     20      │
+///   ╱ ╲    ╱ ╲    ╱ ╲    ╱ ╲    ╱ ╲    ╱ ╲      │
+///  0   1  3   4  7   8  10 11  15 16  18 19 22  ╯
+///
+///  0   1  2   3  4   5  6   7  8   9  10 11 12  ←── leaf indices
+/// ```
+///
+/// Adding another leaf to the above MMR will result in the following MMR:
+///
+/// ```markdown
+///         ──── 14 ───
+///        ╱           ╲
+///       6             13            21
+///      ╱  ╲          ╱  ╲          ╱  ╲
+///     ╱    ╲        ╱    ╲        ╱    ╲
+///    2      5      9      12     17     20    24
+///   ╱ ╲    ╱ ╲    ╱ ╲    ╱ ╲    ╱ ╲    ╱ ╲   ╱ ╲
+///  0   1  3   4  7   8  10 11  15 16  18 19 22 23
+///
+///  0   1  2   3  4   5  6   7  8   9  10 11 12 13
+/// ```
+///
+/// Note how the new leaf with index 13, which by itself is a Merkle tree of
+/// height 0, was merged with its neighboring Merkle tree of equal height,
+/// resulting in a new Merkle tree of height 1 (the nodes 22, 23, and 24).
+///
+/// Adding two more leafs to this MMR results in a single Merkle tree of height
+/// four.
+///
+/// # Counterexample
+///
+/// The following is **not** a valid Merkle Mountain Range:
+///
+/// ```markdown
+///       _                        _
+///      ╱  ╲                     ╱  ╲
+///     ╱    ╲                   ╱    ╲
+///    _      _     _     _     _      _
+///   ╱ ╲    ╱ ╲   ╱ ╲   ╱ ╲   ╱ ╲    ╱ ╲
+///  _   _  _   _ _   _ _   _ _   _  _   _
+/// ```
+///
+/// It has multiple problems: There are two neighboring Merkle trees of equal
+/// height, and there is a Merkle tree that is higher than its left neighbor.
+///
+/// # Terminology
+///
+/// Similar to Merkle trees, Merkle Mountain Ranges have “leafs” and “nodes”.
+/// Any leaf is a node, but the reverse is not true. Note that the indexing
+/// scheme for the nodes does not follow that of a Merkle tree – see
+/// [this][mt_proof] explanation for Merkle tree indexing.
+///
+/// A single Merkle tree's root is one “peak” of the MMR. The number of peaks in
+/// the MMR equals the number of Merkle trees it is made out of. Because of the
+/// 1-to-1 mapping between peaks and Merkle trees, the “peak index” uniquely
+/// identifies any one Merkle tree.
+///
+/// When selecting for a single Merkle tree within a Merkle Mountain Range, the
+/// usual Merkel tree indexing is used to identify nodes. This is called the
+/// “Merkle tree index” and is usually only meaningful in combination with a
+/// peak index, for example like [here][mt_peak_idx].
+///
+/// [mt_proof]: crate::util_types::merkle_tree::MerkleTree::authentication_structure
+/// [mt_peak_idx]: crate::util_types::mmr::shared_basic::leaf_index_to_mt_index_and_peak_index
 pub trait Mmr {
     // constructors cannot be part of the interface since the archival version requires a
     // database which we want the caller to create, and the accumulator does not need a
     // constructor.
```

Comment:
This comment is deprecated, as the archival version of the MMR was moved downstream, to `neptune-core`. This was done because all disk operations were made async such that the archival version could not share a trait (interface) with the accumulator-version.